src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/AbstractDocumentBasedFixAllProvider.cs

```diff
@@ -0,0 +1,232 @@
+                    if (docIdToNewRoot.Count == 0)
+                        continue;
+
+                    // Next, go and insert those all into the solution so all the docs in this particular project point
+                    // at the new trees.  At this point though, the trees have not been postprocessed/cleaned.
+                    foreach (var (docId, newRoot) in docIdToNewRoot)
+                        currentSolution = currentSolution.WithDocumentSyntaxRoot(docId, newRoot);
+
+                    // Next, go and cleanup the trees we inserted.  We do this in bulk so we can benefit from Sharing a
+                    // single compilation across all of them for all the semantic work we need to do. 
+                    //
+                    // Also, once we clean the document, get the text of it and insert that back into the final
+                    // solution.  This way we can release both the original fixed tree, and the cleaned tree (both of
+                    // which can be much more expensive than just text).
+
+                    foreach (var (docId, _) in docIdToNewRoot)
+                    {
+                        var dirtyDocument = currentSolution.GetRequiredDocument(docId);
+                        cleanupTasks.Add(Task.Run(async () =>
+                        {
+                            var cleanedDocument = await PostProcessCodeAction.Instance.PostProcessChangesAsync(dirtyDocument, cancellationToken).ConfigureAwait(false);
+                            var cleanedText = await cleanedDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                            return (dirtyDocument.Id, cleanedText);
+                        }));
+                    }
+
+                    await Task.WhenAll(cleanupTasks).ConfigureAwait(false);
+
+                    foreach (var task in cleanupTasks)
+                    {
+                        var (docId, cleanedText) = await task.ConfigureAwait(false);
+                        currentSolution = currentSolution.WithDocumentText(docId, cleanedText);
+                    }
+                }
+                finally
+                {
+                    progressTracker.ItemCompleted();
+                }
+            }
+
+            return currentSolution;
+        }
+
+        private async Task AddDocumentFixesAsync(
+            FixAllContext fixAllContext,
+            Project project,
+            PooledDictionary<DocumentId, SyntaxNode> docIdToNewRoot)
+        {
+            var cancellationToken = fixAllContext.CancellationToken;
+            var progressTracker = fixAllContext.GetProgressTracker();
+
+            var solution = fixAllContext.Solution;
+
+            // First, get all the diagnostics for this project.
+            progressTracker.Description = string.Format(WorkspaceExtensionsResources.Computing_diagnostics_for_0, project.Name);
```

Comment:
per project we can also give some nice explanations to the user what's going on.  i.e. is it diagcomputation that is expensive, or is it fix'ing the diagnostics that is expensive.